beacon-chain/rpc/eth/events/events.go

```diff
@@ -88,356 +125,407 @@ func (s *Server) StreamEvents(w http.ResponseWriter, r *http.Request) {
+)
+
+// StreamingResponseWriter defines a type that can be used by the eventStreamer.
+// This must be an http.ResponseWriter that supports flushing and hijacking.
+type StreamingResponseWriter interface {
+	http.ResponseWriter
+	http.Flusher
+}
+
+type lazyReader func() io.Reader
+
+type eventStreamer struct {
+	sync.Mutex
+	ctx      context.Context
+	cancel   func()
+	w        StreamingResponseWriter
+	outbox   chan lazyReader
+	finished chan struct{}
+	kaDur    time.Duration
+}
+
+func NewEventStreamer(ctx context.Context, w http.ResponseWriter, buffSize int, kaDur time.Duration) (*eventStreamer, error) {
+	if kaDur == 0 {
+		kaDur = time.Duration(params.BeaconConfig().SecondsPerSlot) * time.Second
 	}
-	return nil
+	f, ok := w.(StreamingResponseWriter)
+	if !ok {
+		return nil, errors.New("beacon node misconfiguration: http stack may not support required response handling features, like flushing")
+	}
+	ctx, cancel := context.WithCancel(ctx)
+	return &eventStreamer{
+		ctx:    ctx,
+		cancel: cancel,
+		w:      f,
+		outbox: make(chan lazyReader, buffSize),
+		kaDur:  kaDur,
+	}, nil
+}
+
+func (es *eventStreamer) Start() {
+	// Set up SSE response headers
+	es.w.Header().Set("Content-Type", api.EventStreamMediaType)
+	es.w.Header().Set("Connection", api.KeepAlive)
+	es.spawnWriteLoop()
 }
 
-func (s *Server) handleStateEvents(ctx context.Context, w http.ResponseWriter, flusher http.Flusher, requestedTopics map[string]bool, event *feed.Event) error {
-	switch event.Type {
-	case statefeed.NewHead:
-		if _, ok := requestedTopics[HeadTopic]; ok {
-			headData, ok := event.Data.(*ethpb.EventHead)
-			if !ok {
-				return write(w, flusher, topicDataMismatch, event.Data, HeadTopic)
+func (es *eventStreamer) Stop() {
+	es.Cleanup(nil)
+}
+
+func newlineReader() io.Reader {
+	return bytes.NewBufferString(":\n\n")
+}
+
+// We main only care about this keep alive value in a real test setups and using a package var allows test
+// to modify it for convenience.
+var keepAliveInterval = time.Duration(12 * time.Second)
+
+func (es *eventStreamer) spawnWriteLoop() {
+	go func() {
+		var err error
+		kaT := time.NewTimer(es.kaDur)
+		defer func() {
+			if !kaT.Stop() {
+				<-kaT.C
 			}
-			head := &structs.HeadEvent{
-				Slot:                      fmt.Sprintf("%d", headData.Slot),
-				Block:                     hexutil.Encode(headData.Block),
-				State:                     hexutil.Encode(headData.State),
-				EpochTransition:           headData.EpochTransition,
-				ExecutionOptimistic:       headData.ExecutionOptimistic,
-				PreviousDutyDependentRoot: hexutil.Encode(headData.PreviousDutyDependentRoot),
-				CurrentDutyDependentRoot:  hexutil.Encode(headData.CurrentDutyDependentRoot),
+		}()
+		defer es.Cleanup(err)
+		for {
+			select {
+			case <-es.ctx.Done():
+				return
+			case <-kaT.C:
+				err = es.writeOutbox(nil)
+				if err != nil {
+					return
+				}
+				// The timer has already fired here, so a call to Reset is safe.
```

Comment:
The reason for this comment is that < go 1.23 the `Reset` docs used to warn: 
`For a Timer created with NewTimer, Reset should be invoked only on stopped or expired timers with drained channels.`

So I meant to convey that when `Reset` is called below, it's ok that the code in this case doesn't call `Stop` first, because we know the timer has fired. That's why in the other case (read from outbox) you see this comment: 

`We don't know if the timer fired concurrently to this case being ready, so we need to check the return of Stop and drain the timer channel if it fired.`

I'm rewriting the comment to make it more clear.

Looks like in go 1.23 they cleaned all this up (and timers without any references can be garbage collected, even if they the code forgot to call `Stop`, yay!